src/server/webkit/wkPage.ts

```diff
@@ -1160,3 +1195,62 @@ function isLoadedSecurely(url: string, timing: network.ResourceTiming) {
     return true;
   } catch (_) {}
 }
+
+
+class WKNavigation {
+  private readonly _page: WKPage;
+  // loaderId is a navigation id which is unique within browser and persists during cross-process navigation.
+  private readonly _loaderId: string;
+  readonly _originalRequestId: Protocol.Network.RequestId;
+  private readonly _frameId: string;
+  private _responseReceived: boolean = false;
+  private _didCheckFirstRequestInProvisionalPage: boolean = false;
+  private _provisionalPageRequestId: string | undefined;
+
+  constructor(page: WKPage, event: Protocol.Network.requestWillBeSentPayload) {
+    this._page = page;
+    this._loaderId = event.loaderId;
+    this._originalRequestId = event.requestId;
+    this._frameId = event.frameId;
+  }
+
+  shouldIgnoreFrameStoppedLoading(frameId: string): boolean {
+    if (frameId !== this._frameId)
+      return false;
+    // Navigation in the original frame is canceled and actually continues in the
+    // provisional page, so we ignore failure events from the original page.
+    return this._page.hasProvisionalPage();
+  }
+
+  shouldIgnoreLoadingFailedEvent(event: Protocol.Network.loadingFailedPayload): boolean {
+    if (event.requestId !== this._originalRequestId)
+      return false;
+    if (!event.canceled)
+      return false;
+    // Navigation in the original frame is canceled and actually continues in the
+    // provisional page, so we ignore failure events from the original page.
+    return this._page.hasProvisionalPage();
+  }
+
+  checkIfSameNavigationInNewProcess(event: Protocol.Network.requestWillBeSentPayload): boolean {
+    if (this._didCheckFirstRequestInProvisionalPage)
+      return false;
+    this._didCheckFirstRequestInProvisionalPage = true;
+    if (event.loaderId !== this._loaderId)
+      return false;
+    this._provisionalPageRequestId = event.requestId;
+    return true;
+  }
+
+  checkIfDuplicateResponseEvent(event: Protocol.Network.responseReceivedPayload): boolean {
+    if (this._originalRequestId === event.requestId || this._provisionalPageRequestId === event.requestId) {
```

Comment:
Can we ignore responses from the old web process only? This ensures that we get correct response from the new web process, in case WebKit fixes more security holes in the old web process and send bogus response from there (if any).

Reply:
Coincidentally that works because we generate loadingFinished event in WebKit ourselves when 204 response comes: https://github.com/microsoft/playwright/blob/45b365d95833e0ab61bd49e0b98293542fbd12c0/src/server/webkit/wkPage.ts#L1023-L1030

With network failures we will receive `Playwright.provisionalLoadFailed` and that will abort the navigation, so we are good in that case too. Added a test.

Reply:
At the moment response notification comes from old process we don't know if it's gonna be a cross-process or same process navigation. In case of same-process navigation we have to dispatch messages from the old process. Unless we introduce a separate notification that the navigation turned into cross-process and buffer up some messages from old web process before the notification arrives we cannot safely ignore responses from the old web process  .